### PR TITLE
HUBS-1603 Plan phase hangs when VDB resource no longer exists

### PR DESCRIPTION
**Pull request by kishlay-delphix**
# Problem:

When the VDB was deleted from the engine and terraform plan was being executed, the drift detection mechanism in the provider was not working as expected.Even though the VDB was deleted from the remote the state file reference wasn't getting cleaned up resulting in a hang like behaviour.
# Solution:

The drift mechanism was fixed and terraform plan was executing as expected.Also we added extra checks to ensure there is no false override of the state file.
# Testing

Tested VDB deletion flow and terraform commands

### CODE REVIEW

File: internal/provider/resource_vdb.go [hunk 1]
```diff
@@ -1213,14 +1213,25 @@ func resourceVdbRead(ctx context.Context, d *schema.ResourceData, meta interface
...
-		return diag.Errorf("[NOT OK] Error in polling vdb")
+		_, diags, _ := PollForObjectDeletion(func() (interface{}, *http.Response, error) {
+			return client.VDBsApi.GetVdbById(ctx, vdbId).Execute()
+		})
+		// This would imply error in poll for deletion so we just log and exit.
+		if diags != nil{
+			ErrorLog.Printf("[NOT OK] Error in polling vdb")
+		}
+		// diags will be nill in case of successful poll for deletion logic aka 404
+		if diags == nil && statusCode==404{
```
Comment (sunnyx4) [inline internal/provider/resource_vdb.go:1229]: statusCode needs to be set at line 1221?

File: internal/provider/resource_vdb.go [hunk 2]
```diff
@@ -1213,14 +1213,25 @@ func resourceVdbRead(ctx context.Context, d *schema.ResourceData, meta interface
...
 	if diags != nil {
-		ErrorLog.Printf("Error reading the VDB, removing from state.")
-		d.SetId("")
-		return diag.Errorf("[NOT OK] Error in polling vdb")
+		_, diags, _ := PollForObjectDeletion(func() (interface{}, *http.Response, error) {
+			return client.VDBsApi.GetVdbById(ctx, vdbId).Execute()
+		})
+		// This would imply error in poll for deletion so we just log and exit.
+		if diags != nil{
+			ErrorLog.Printf("[NOT OK] Error in polling vdb")
```
Comment (sunnyx4) [inline internal/provider/resource_vdb.go:1226]: Update the error message to have more information. In which poll? what is the error? etc

File: internal/provider/resource_vdb.go [hunk 3]
```diff
@@ -1213,14 +1213,25 @@ func resourceVdbRead(ctx context.Context, d *schema.ResourceData, meta interface
...
+		}
+		// diags will be nill in case of successful poll for deletion logic aka 404
+		if diags == nil && statusCode==404{
+			ErrorLog.Printf("Error reading the VDB %s, removing from state.",vdbId)
```
Comment (sunnyx4) [inline internal/provider/resource_vdb.go:1230]: This error message is misleading. It should say received 404 hence removing from state

File: internal/provider/utility.go [hunk 4]
```diff
@@ -62,32 +62,32 @@ func ResponseBodyToString(body io.ReadCloser) (string, error) {
...
 
-func PollForObjectDeletion(apiCall func() (interface{}, *http.Response, error)) (interface{}, diag.Diagnostics) {
+func PollForObjectDeletion(apiCall func() (interface{}, *http.Response, error)) (interface{}, diag.Diagnostics, int) {
 	// Function to check if an object does not exist in the Delphix estate.
 	return PollForStatusCode(apiCall, http.StatusNotFound, 10)
 }
 
 // poll counter is the retry counter for which an api call should be retried.
-func PollForStatusCode(apiCall func() (interface{}, *http.Response, error), statusCode int, maxRetry int) (interface{}, diag.Diagnostics) {
+func PollForStatusCode(apiCall func() (interface{}, *http.Response, error), statusCode int, maxRetry int) (interface{}, diag.Diagnostics,int) {
```
Comment (sunnyx4) [inline internal/provider/utility.go:76]: General comment: Are you using any formatting tool? something like gofmt? For example, there should be space between diag.Diagnostics, and int but it is not. This spacing issues are present at other locations as well

File: internal/provider/resource_environment.go [hunk 5]
```diff
@@ -385,7 +385,7 @@ func resourceEnvironmentRead(ctx context.Context, d *schema.ResourceData, meta i
 	client := meta.(*apiClient).client
 	envId := d.Id()
 
-	apiRes, diags := PollForObjectExistence(func() (interface{}, *http.Response, error) {
+	apiRes, diags, _ := PollForObjectExistence(func() (interface{}, *http.Response, error) {
```
Comment (sunnyx4) [inline internal/provider/resource_environment.go:388]: Wouldn't we see the same bug behavior in the Environment resource?

**Review by sunnyx4 — CHANGES_REQUESTED** [review]

File: internal/provider/resource_vdb.go (hunk 1 above)
Comment (sunnyx4) [reply internal/provider/resource_vdb.go:1229]: statusCode is supposed to be set at 1221 right? instead you are using _ in third return value

Comment (kishlay-delphix) [reply internal/provider/resource_vdb.go:1229]: We don't need to set status code for `PollForObjectDeletion` as it polls for 404 and it being successful means that it returned 404 so we are checking for statusCode of only `PollForObjectExistence` line 1216

Comment (sunnyx4) [reply internal/provider/resource_vdb.go:1229]: Yes, but if you are not setting statusCode at 1221, how would your IF condition at line 1229 will match for "404" value?

Comment (kishlay-delphix) [reply internal/provider/resource_vdb.go:1229]: We are setting it on line 1216

**Comment by nick-mathison** [comment]
Unsure of the exact release process here, but we'll need to update the version to `1.0.1` before we tag a release.

**Pull request closed without merging**